models/issue_comment.go

```diff
@@ -1016,3 +1054,147 @@ func UpdateCommentsMigrationsByType(tp structs.GitServiceType, originalAuthorID
 		})
 	return err
 }
+
+// CreatePushPullComment create push code to pull base commend
+func CreatePushPullComment(pusher *User, pr *PullRequest, oldCommitID, newCommitID string) (comment *Comment, err error) {
+	if pr.HasMerged || oldCommitID == "" || newCommitID == "" {
+		return nil, nil
+	}
+
+	ops := &CreateCommentOptions{
+		Type: CommentTypePullPush,
+		Doer: pusher,
+		Repo: pr.BaseRepo,
+	}
+
+	var commitIDs []string
+	messages := ""
+
+	var isForcePush bool
+	commitIDs, messages, isForcePush, err = getCommitIDsFromRepo(pr.BaseRepo, oldCommitID, newCommitID, pr.BaseBranch)
+	if err != nil {
+		return nil, err
+	}
+
+	ops.LineNum = int64(len(commitIDs))
+	ops.Issue = pr.Issue
+	ops.IsForcePush = isForcePush
+
+	commitIDlist := ""
+
+	for _, commitID := range commitIDs {
+		commitIDlist = commitID + ":" + commitIDlist
+	}
+
+	ops.Content = commitIDlist[0 : len(commitIDlist)-1]
+
+	comment, err = CreateComment(ops)
+	if err != nil {
+		return
+	}
+
+	// Prepare the contents for notify
+	prmessages := "@" + ops.Doer.Name
+	if ops.IsForcePush {
+		prmessages += " force-pushed the " + pr.HeadBranch + " branch from " + oldCommitID + " to " + newCommitID
+	} else {
+		if ops.LineNum == 1 {
+			prmessages += fmt.Sprintf(" pushed 1 commit to %s:  \n ", pr.HeadBranch)
+		} else {
+			prmessages += fmt.Sprintf(" pushed %d commits to %s:  \n ", ops.LineNum, pr.HeadBranch)
+		}
+		prmessages += messages
+	}
+
+	comment.Content = prmessages
+
+	return
+}
+
+// getCommitsFromRepo get commit IDs from repo in betwern oldCommitID and newCommitID
+// isForcePush will be true if oldCommit isn't on the branch
+// Commit on baseBranch will skip
+func getCommitIDsFromRepo(repo *Repository, oldCommitID, newCommitID, baseBranch string) (commitIDs []string, messages string, isForcePush bool, err error) {
+	repoPath := repo.RepoPath()
+	gitRepo, err := git.OpenRepository(repoPath)
+	if err != nil {
+		return nil, "", false, err
+	}
+	defer gitRepo.Close()
+
+	oldCommit, err := gitRepo.GetCommit(oldCommitID)
+	if err != nil {
+		return nil, "", false, err
+	}
+
+	oldCommitBranch, err := oldCommit.GetBranchName()
+	if err != nil {
+		return nil, "", false, err
+	}
+
+	if oldCommitBranch == "undefined" {
+		commitIDs = make([]string, 2)
+		commitIDs[1] = oldCommitID
+		commitIDs[0] = newCommitID
+
+		return commitIDs, "", true, err
+	}
+
+	newCommit, err := gitRepo.GetCommit(newCommitID)
+	if err != nil {
+		return nil, "", false, err
+	}
+
+	var commits *list.List
+	commits, err = newCommit.CommitsBeforeUntil(oldCommitID)
+	if err != nil {
+		return nil, "", false, err
+	}
+
+	commitIDs = make([]string, 0, commits.Len())
+	messages = ""
+
+	for e := commits.Front(); e != nil; e = e.Next() {
+		commit := e.Value.(*git.Commit)
+		commitBranch, err := commit.GetBranchName()
```

Comment:
Because the commits on base branch added by merge base branch to head branch will also be found on this list. but it's not necessary to show them on view. so skip them.
view on github.
![image](https://user-images.githubusercontent.com/25342410/80174792-56b80080-8626-11ea-89c5-8a6a53fddd78.png)
commit will get:
![image](https://user-images.githubusercontent.com/25342410/80175099-3dfc1a80-8627-11ea-8c9b-bf5aa498bbfa.png)

Reply:
Couldn't there be unrelated branches in the list? I could created as many branches as I want on each commit, and if I don't commit to that branch the commit will be in the same line as the PR...

Reply:
Not yet, only remove commits on the base brach which is same as what show in commits view in pull request. Thanks.